sql/src/main/java/io/crate/execution/engine/indexing/ShardedRequests.java

```diff
@@ -48,46 +48,65 @@
     /**
      * @param requestFactory function to create a request
      */
-    public ShardedRequests(Function<ShardId, TReq> requestFactory) {
+    ShardedRequests(Function<ShardId, TReq> requestFactory) {
         this.requestFactory = requestFactory;
     }
 
-    public void add(TItem item, String indexName, String routing, @Nullable BytesRef sourceUri) {
-        List<ItemAndRoutingAndSourceUri<TItem>> items = itemsByMissingIndex.computeIfAbsent(indexName, k -> new ArrayList<>());
-        items.add(new ItemAndRoutingAndSourceUri<>(item, routing, sourceUri));
+    public void add(TItem item,
+                    String indexName,
+                    String routing,
+                    @Nullable BytesRef sourceUri,
```

Comment:
Could we require the `RowSourceInfo` here already to make it clear that the two things belong together and are both either present or null?